lib/Dialect/Synth/Transforms/CutRewriter.cpp

```diff
@@ -0,0 +1,564 @@
+
+  if (isTrivialCut()) {
+    // For a trivial cut, a truth table is simply the identity function.
+    // 0 -> 0, 1 -> 1
+    truthTable = BinaryTruthTable(1, 1, {llvm::APInt(2, 2)});
+    return *truthTable;
+  }
+
+  int64_t numInputs = getInputSize();
+  int64_t numOutputs = getOutputSize();
+  assert(numInputs < 20 && "Truth table is too large");
+
+  // Simulate the IR.
+  uint32_t tableSize = 1 << numInputs;
+  DenseMap<Value, APInt> eval;
+  for (uint32_t i = 0; i < numInputs; ++i) {
+    APInt value(tableSize, 0);
+    for (uint32_t j = 0; j < tableSize; ++j) {
+      // Make sure the order of the bits is correct.
+      value.setBitVal(j, (j >> i) & 1);
+    }
+    // Set the input value for the truth table
+    eval[inputs[i]] = std::move(value);
+  }
+
+  // Simulate the operations in the cut
+  for (auto *op : operations)
+    simulateLogicOp(op, eval);
+
+  // Extract the truth table from the root operation
+  auto rootResults = getRoot()->getResults();
+  BinaryTruthTable result(numInputs, numOutputs);
+  assert(numOutputs == 1 &&
+         "Multiple outputs are not supported yet, must be rejected earlier");
+  result.table = eval.at(rootResults[0]);
+  truthTable = std::move(result);
+  return *truthTable;
+}
+
+static Cut getAsTrivialCut(mlir::Value value) {
+  // Create a cut with a single root operation
+  Cut cut;
+  // There is no input for the primary input cut.
+  cut.inputs.insert(value);
+
+  return cut;
+}
+
+Cut Cut::mergeWith(const Cut &other, Operation *root) const {
+  // Create a new cut that combines this cut and the other cut
+  Cut newCut;
+  // Topological sort the operations in the new cut.
+  // TODO: Merge-sort `operations` and `other.operations` by operation index
+  // (since it's already topo-sorted, we can use a simple merge).
+  std::function<void(Operation *)> populateOperations = [&](Operation *op) {
+    // If the operation is already in the cut, skip it
+    if (newCut.operations.contains(op))
+      return;
+
+    // Add its operands to the worklist
+    for (auto value : op->getOperands()) {
+      if (isAlwaysCutInput(value))
+        continue;
+
+      // If the value is in *both* cuts inputs, it is an input. So skip
+      // it.
+      bool isInput = inputs.contains(value);
+      bool isOtherInput = other.inputs.contains(value);
+      // If the value is in this cut inputs, it is an input. So skip it
+      if (isInput && isOtherInput)
+        continue;
+
+      auto *defOp = value.getDefiningOp();
+
+      assert(defOp && "Value must have a defining operation since block"
+                      "arguments are treated as inputs");
+
+      // Otherwise, check if the operation is in the other cut.
+
+      if (isInput) {
+        if (!other.operations.contains(defOp)) // op is in the other cut.
+          continue;
+      }
+      if (isOtherInput) {
+        if (!operations.contains(defOp)) // op is in this cut.
+          continue;
+      }
+
+      populateOperations(defOp);
```

Comment:
Could compact this slightly:
```suggestion
      // Otherwise, check if the operation is in the other cut.
      auto *defOp = value.getDefiningOp();
      assert(defOp && "Value must have a defining operation since block"
                      "arguments are treated as inputs");
      if (isInput)
        if (!other.operations.contains(defOp)) // op is in the other cut.
          continue;
      if (isOtherInput)
        if (!operations.contains(defOp)) // op is in this cut.
          continue;
      populateOperations(defOp);
```